mono/mini/mini-arm64.c

```diff
@@ -3678,76 +3682,94 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 
 			/* sreg2 is the value, sreg3 is the comparand */
 			buf [0] = code;
-			arm_ldaxrw (code, ARMREG_IP0, sreg1);
+			arm_ldxrw (code, ARMREG_IP0, sreg1);
 			arm_cmpw (code, ARMREG_IP0, ins->sreg3);
 			buf [1] = code;
 			arm_bcc (code, ARMCOND_NE, 0);
 			arm_stlxrw (code, ARMREG_IP1, sreg2, sreg1);
 			arm_cbnzw (code, ARMREG_IP1, buf [0]);
 			arm_patch_rel (buf [1], code, MONO_R_ARM64_BCC);
 
+			arm_dmb (code, 0);
 			arm_movx (code, dreg, ARMREG_IP0);
 			break;
 		}
 		case OP_ATOMIC_CAS_I8: {
 			guint8 *buf [16];
 
 			buf [0] = code;
-			arm_ldaxrx (code, ARMREG_IP0, sreg1);
+			arm_ldxrx (code, ARMREG_IP0, sreg1);
 			arm_cmpx (code, ARMREG_IP0, ins->sreg3);
 			buf [1] = code;
 			arm_bcc (code, ARMCOND_NE, 0);
 			arm_stlxrx (code, ARMREG_IP1, sreg2, sreg1);
 			arm_cbnzw (code, ARMREG_IP1, buf [0]);
 			arm_patch_rel (buf [1], code, MONO_R_ARM64_BCC);
 
+			arm_dmb (code, 0);
 			arm_movx (code, dreg, ARMREG_IP0);
 			break;
 		}
 		case OP_ATOMIC_LOAD_I1: {
 			code = emit_addx_imm (code, ARMREG_LR, ins->inst_basereg, ins->inst_offset);
+			if (ins->backend.memory_barrier_kind == MONO_MEMORY_BARRIER_SEQ)
```

Comment:
This looks wrong.

We emit a full fence and then using ldarb?

It should be one or the other.  dmb + ldrb or ldarb.